XMonad/Doc/Extending.hs

```diff
@@ -237,124 +230,161 @@ yourself.
->             , ((modm, xK_F3 ), shellPrompt  def)
->             ]
-
-There are much simpler ways to accomplish this, however, if you are
-willing to use an extension module to help you configure your keys.
-For instance, "XMonad.Util.EZConfig" and "XMonad.Util.CustomKeys" both
-provide useful functions for editing your key bindings; "XMonad.Util.EZConfig" even lets you use emacs-style keybinding descriptions like \"M-C-<F12>\".
+> main :: IO ()
+> main = xmonad $ def { keys = myKeys }
+>  where
+>   myKeys :: XConfig l -> Map (ButtonMask, KeySym) (X ())
+>   myKeys _conf = Map.fromList
+>     [ ((mod1Mask, xK_m        ), spawn "echo 'Hi, mom!' | dzen2 -p 4")
+>     , ((mod1Mask, xK_BackSpace), spawn "xterm")
+>     ]
+
+Now, obviously we don't want to do that; we only want to add to existing
+bindings (or, perhaps, override some of them with our own).  Let's break
+@myKeys@ down a little.  You can think of the type signature like this:
+
+>    myKeys :: UserConfig -> Map KeyPress Action
+
+It takes some user config and, from that, produces a map that associates
+certain keypresses with actions to execute.  Because we're setting @keys
+= myKeys@ above, this is also the type of 'XMonad.Core.keys'.
+
+This means that, as a first guess, the type signature of our version of
+'XMonad.Util.EZConfig.additionalKeys' might look like
+
+> myAdditionalKeys :: XConfig l
+>                     -- ^ Base config with xmonad's default keybindings
+>                  -> (XConfig l -> Map (ButtonMask, KeySym) (X ()))
+>                     -- ^ User supplied keybindings
+>                  -> XConfig l
+>                     -- ^ Resulting config with everything merged together
+
+However, even assuming a correct implementation, using this is not very
+ergonomic:
+
+> main = xmonad $ def
+>  `myAdditionalKeys`
+>    (\ _conf -> Map.fromList
+>      [ ((mod1Mask, xK_m        ), spawn "echo 'Hi, mom!' | dzen2 -p 4")
+>      , ((mod1Mask, xK_BackSpace), spawn "xterm")
+>      ])
+
+Having to specify a lambda with parentheses and call
+'Data.Map.Strict.fromList' does not make for a good user experience.
+Since one /always/ has to call that function anyways, we may well just
+accept a list from the user and transform it to a map ourselves.  As an
+additional simplification, how about we don't care about the config
+argument at all and simply ask the user for a list?  The resulting
+signature
+
+> myAdditionalKeys :: XConfig l
+>                  -> [(ButtonMask, KeySym), (X ())]
+>                  -> XConfig l
+
+looks exactly like what we want!  Now that we know what kind of data
+structure—that is, maps—we are dealing with, the implementation of this
+function simply merges the two together, preferring the user config to
+xmonad's defaults in case of any conflicts.  Thankfully, someone else
+has already done the hard work and written the merging function for us;
+it's called 'Data.Map.Strict.union'.
+
+What's left is essentially playing type tetris:
```

Comment:
It's all very well to call it that informally, but it strikes me that (a) it's a bad idea in more formal documentation (b) it may run into copyright / licensing issues.

Reply:
This whole section is somewhat informal though, so "type tetris" doesn't stand out too much.
Could you elaborate a bit about the copyright/licensing issue? Are you suggesting that one cannot use the word tetris in a documentation? \[citation needed\] :-)

Reply:
Not sure I can point to specific law, but trademarks and in particular registered trademarks have to be enforced to not be lost in the U.S. and probably other countries. See "Kleenex" for one that's in danger of being turned over, since hardly anyone calls them "tissues". Which means we risk being contacted by a lawyer if we use it.

Reply:
According to https://desiree47.wordpress.com/2008/12/07/tetris-and-intellectual-property-rights-trademark/amp/:

> Trademark rights prevent outside vendors from using identical or “confusingly similar” marks on their merchandise to sell their products (USPTO).
> […]
> The Tetris Company has obtained registered trademarks for the use of the word TETRIS and the Tetris logo in the industries of:
>
> “games and playthings; namely action figures, bagatelle games; video games and video game cartridges; electronic game equipment for playing video games; electronic game programs; electronic toys with a clock or timepiece feature; hand held electronic toys; non-electronic hand held games; player operated electronic controllers for electronic video game machines; toy figures mountable on pencils and pens,”
>
> “video game cartridges and software for playing video games,”
>
> “clothing, namely shirts, hats and t-shirts,”and “hand held video games”.

We're not using it for marketing/sales purposes, and we're not in any of these industries, so I'd say it should be fine, but IANAL.

Reply:
So, at this point, what's the general feeling here: keep/delete?

Reply:
My vote is "keep". You wouldn't even be the first to use "type tetris".